### PR TITLE
fix(undo): operation-aware json_patch cell undo via read-modify-write (#425 part 1 of 2)

### LINKED ISSUE #425: Hot-exit does not persist redo/undone state (futureStack) — undo of a saved edit is lost on reopen
Split out from PR #424 review (CodeRabbit, `src/core/undo-history.ts:258`) as a tracked follow-up. **Pre-existing on `main`** — not introduced by the web-edit PR — and affects **both desktop and web**, so it's out of scope for #424 and deserves its own change.

## Problem
`ModificationTracker.serialize()` persists only `{ timeline, checkpointIndex }`:

```ts
// src/core/undo-history.ts
serialize(): Uint8Array {
  const payload = { timeline: this.timeline, checkpointIndex: this.checkpointIndex };
  ...
}
```

But `stepBack()` (undo) and `rollbackToCheckpoint()` move entries into `futureStack` (the redo stack), which is **not serialized**. Hot-exit restore opens the last *saved* DB bytes and replays only `getUncommittedEntries()` (`timeline.slice(checkpointIndex)`).

## Consequence
Repro:
1. Make an edit, **save** (checkpoint advances past the edit).
2. **Undo** that saved edit (entry moves to `futureStack`; document is now dirty).
3. Hot-exit (close VS Code with the dirty doc), then reopen from the backup.

Expected: the database reflects the **undone** state (edit removed).
Actual: restore opens the saved bytes (which still contain the edit) and replays no forward entries, so the **undone state is lost** — and the redo stack is gone entirely.

## Possible approaches (design decision — not prescribing)
- **Persist `futureStack` + `futureStackSizes`** and reconstruct the cursor on restore, then replay the net state (requires representing the undo, e.g. applying inverse ops or re-deriving live state).
- **Snapshot the live DB bytes** for the hot-exit backup instead of replaying forward entries from the saved image (exact, simpler to reason about, larger backups).

CodeRabbit tagged this a "Heavy lift"; flagging for a dedicated PR with its own tests (desktop + web hot-exit round-trips covering undo-of-saved-edit).

Refs: PR #424 review thread on `src/core/undo-history.ts`.

### PR DESCRIPTION
**Pull request by zknpr**
## Summary

Part **1 of 2** for #425. Fixes the undo-fidelity sub-problem: **undo of a JSON cell edit no longer clobbers a concurrent change to a *different* key of the same cell.**

Undo previously replaced the whole cell with the recorded prior document (`SET col = priorValue`). If a different key changed between the tracked edit and the undo — e.g. the cell is open in a VFS editor tab while it's also edited inline in the grid — that concurrent change was lost. P2 (data fidelity; single-pane/sequential undo was already correct).

## Approach — read-modify-write (RMW)

A pure helper `computeJsonPatchUndo(current, forwardPatch, prior)` reads the cell's **current** value, restores only the keys the forward merge-patch touched (drawing values from the recorded prior), and writes the full object back; it value-replaces when `current`/`prior`/`forwardPatch` is not a JSON object. All JSON reasoning happens in JS — **no `json_patch()` SQL on the undo path** — which eliminates RFC 7396's `null`=delete ambiguity and `json_patch()`'s silent-wrong behavior on non-object inputs.

This supersedes the closed PR #426, whose inverse-merge-patch approach was architecturally lossy and was closed with two unresolved P2 edges (nested explicit nulls; non-object current cell). RMW closes that entire class. It reuses #426's test scenarios, the `isObject` arrays-are-atomic fix (RFC 7396), and the `operation`/`newValue` recording plumbing already on `main`.

## Behavior (acceptance scenarios)

- Edited key restored; **concurrent sibling key preserved**.
- A wholly-added object key is removed entirely (no leftover `{}`).
- Explicit `null` restored exactly (including nested) while preserving concurrent siblings.
- Current cell became non-object (SQL `NULL` / scalar / array / non-JSON text) → value-replace to prior.
- Forward patch was a scalar/array whole-doc replace, or prior was SQL `NULL` → value-replace.
- Batch undo: per-cell RMW, atomic (WASM `SAVEPOINT` / native single `execBatch`).

## Verification

- `npx tsc --noEmit` clean · `node scripts/build.mjs` OK (all four bundles + `sqlite3.wasm`) · full unit suite **375 / 375**.
- **Revert-proof:** reverting the engine code fails exactly the 6 RMW tests (native ×3, WASM `s1`/`s4`/`s9`); restoring → green.

## Scope

- **No version bump and no CHANGELOG change.** `1.5.1` ships with **Part 2** (futureStack hot-exit persistence), the PR that closes #425. This PR references but does **not** close #425.

Design + plan: `docs/superpowers/specs/2026-06-03-json-patch-undo-rmw-design.md`, `docs/superpowers/plans/2026-06-03-json-patch-undo-rmw.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Improved JSON cell-edit undo: undos now restore only the keys touched by the edit, preserving concurrent sibling key changes and handling edge cases to prevent data loss.

* **Tests**
  * Added extensive tests validating undo behavior across concurrent sibling updates, nested objects, explicit nulls, precision edge cases, and atomic multi-cell batch undos.

[Review Change Stack](https://app.coderabbit.ai/change-stack/zknpr/SQLite-Explorer/pull/427?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by zknpr** [comment]
## Review feedback addressed — all 9 findings

Across 4 commits (`a389f84`, `2e52b44`, `3351523`, `44f2b72`). Full unit suite **384/384**, `tsc --noEmit` clean, build OK.

| Finding | Resolution | Commit |
|---|---|---|
| **CR3** (Major, `__proto__`) | `restoreInto` builds null-prototype containers + `hasOwnProperty` guards, so `__proto__`/`constructor` are treated as data. Test asserts exact serialization **and** `Object.prototype` is untouched. | `a389f84` |
| **CX1** (P2, nested non-object) | Falls back to the full prior subtree when the current nested value is no longer an object — no longer drops untouched prior siblings (e.g. `owner`). | `a389f84` |
| **CX3** (P2, number precision) | `hasPrecisionRiskyInteger` → value-replace for cells containing integers > 2⁵³, keeping them **byte-exact** (RMW reparse was a regression vs the old verbatim restore). Engine test asserts `9007199254740993` survives undo unrounded. | `a389f84` |
| **G1** (Med, WASM) | Manual `SAVEPOINT` + per-cell loop → single `updateCellBatch` (already atomic). | `2e52b44` |
| **G2** (Med, native) | Dropped the `hasJsonPatchUndo` branch; one unified compute/write path. | `3351523` |
| **CR4** (Trivial, native) | Batch the per-cell SELECT reads into a single `queryBatch` round-trip (order-preserving). | `3351523` |
| **CX2** (P2, native atomicity) | Per-connection promise-chain operation lock so a json_patch undo's read/compute/write cannot interleave with a concurrent same-cell write; internal facade→facade calls use the raw (unlocked) impl to avoid deadlock. Regression test asserts a concurrent `updateCell` cannot write between undo's read and write. | `3351523` |
| **CR1/CR2** (Minor) | MD018/MD040 markdown-lint fixes in the design spec. | `44f2b72` |

Note: automated reviewers run on PR open only, so these fixes won't trigger a re-review pass.

File: src/core/json-utils.ts [hunk 1]
```diff
@@ -108,6 +108,114 @@ export function applyMergePatch(target: unknown, patch: unknown, depth = 0): unk
...
+    return result;
+}
+
+function hasPrecisionRiskyInteger(raw: unknown): boolean {
+    if (typeof raw !== 'string') return false;
+    // Neutralize string literals so only structural JSON number tokens are
+    // inspected for integer runs that JavaScript cannot represent exactly.
+    const structural = raw.replace(/"(?:\\.|[^"\\])*"/g, '""');
+    const digitRuns = structural.match(/\d{16,}/g);
+    return !!digitRuns && digitRuns.some(run => !Number.isSafeInteger(Number(run)));
```
[the remark this answers is not in the document]
Comment (zknpr) [reply src/core/json-utils.ts]: Fixed in 71cb539. `hasPrecisionRiskyInteger` → `hasPrecisionRiskyNumber`, which now also rejects any token that parses to a non-finite number (covers `1e999` → `Infinity`), so those cells value-replace and stay byte-exact. Engine test asserts a `{"huge":1e999,...}` cell survives a json_patch undo with `huge` unchanged (not `null`).

File: src/core/engine/wasm/WasmDatabaseEngine.ts [hunk 2]
```diff
@@ -282,19 +282,67 @@ export class WasmDatabaseEngine implements DatabaseOperations {
...
+  private async computeUndoValue(
+    table: string,
+    rowId: RecordId,
+    column: string,
+    priorValue: CellValue | undefined,
+    newValue: CellValue | undefined,
+    operation: ModificationEntry['operation']
+  ): Promise<CellValue> {
+    if (operation === 'json_patch') {
+      const currentValue = await this.readCellValue(table, rowId, column);
```
[the remark this answers is not in the document]
Comment (zknpr) [reply src/core/engine/wasm/WasmDatabaseEngine.ts:327]: Fixed in 96d0f89. Correct — the native CX2 lock didn't cover the WASM in-process path. Extracted the per-connection promise-chain lock into a shared `serializeOperations()` (Proxy over DatabaseOperations; passes `engineKind`/dispose through; deadlock-free since internal facade calls bypass it) and applied it to **both** facades: native (replacing the inline lock) and the WASM in-process facade in `workerFactory`. The worker-thread WASM path is one-RPC-per-op, so it's already atomic and isn't wrapped. Added a `serializeOperations` unit test (FIFO + error-release) alongside the existing native serialization test.

File: src/core/json-utils.ts [hunk 3]
```diff
@@ -108,6 +108,119 @@ export function applyMergePatch(target: unknown, patch: unknown, depth = 0): unk
...
+function hasPrecisionRiskyNumber(raw: unknown): boolean {
+    if (typeof raw !== 'string') return false;
+    // Neutralize string literals so only structural JSON number tokens are inspected.
+    const structural = raw.replace(/"(?:\\.|[^"\\])*"/g, '""');
+    const tokens = structural.match(/-?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?/g);
+    if (!tokens) return false;
+    return tokens.some(tok => {
+        const n = Number(tok);
+        if (!Number.isFinite(n)) return true;
+        if (/^-?\d+$/.test(tok)) return !Number.isSafeInteger(n);
```
[the remark this answers is not in the document]
Comment (zknpr) [reply src/core/json-utils.ts]: Fixed in e0e6b3b. Replaced the integer-only check with a uniform round-trip test — a token is risky if non-finite or `JSON.stringify(Number(tok)) !== tok` — so large integers, exponent overflow (`1e999`), and high-precision decimals (`0.1234567890123456789012345`) all fall back to byte-exact value-replace in one rule. Pure + engine tests assert the decimal survives undo unrounded. Full suite 390/390 (no over-flag regression — canonical numbers still take the surgical path).
